xsnap/documentation/lockdown.md

```diff
@@ -0,0 +1,238 @@
+	h.r.i = createObservable(null, "i");
+	h.r.t = createObservable(null, "t");
+	h.d = createObservable(null, "d");
+	
+When hardening `h`, the script prints `hardenit`!	
+	
+	lockdown();
+	harden(h);
+	print(order);
+
+
+
+## purify(o)
+
+The experimental `purify` function checks the immutability of the passed object.
+
+- The `purify` function checks the passed object, then checks its prototypes and all its properties.
+- The `purify` function does not use standard internal methods so JS code is never executed while checking.
+- The `purify` function returns a string: a `'\n'` separated list of all mutable property paths, or the empty string if all properties are immutable.
+
+The behavior of the `purify` function is expected to change.
+
+For instance
+
+	lockdown();
+	const o = {
+		foo:88,
+	};
+	print(purify(o));
+
+prints
+
+	.foo
+
+But
+
+	lockdown();
+	const p = harden({
+		foo:88,
+	});
+	print(purify(p));
+
+prints nothing.
+
+The `purify ` function checks closures in function environments.
+
+	lockdown();
+	let v = 0;
+	const o = {
+		oops(i) {
+			return v + i;
+		},
+	};
+	harden(o);
+	print(purify(o));
+
+prints
+	
+	.oops[[Environment]].v
+
+since `v` is variable.
+
+But 
+
+	lockdown();
+	const c = 0;
+	const p = {
+		oops(i) {
+			return c + i;
+		},
+	};
+	harden(p);
+	print(purify(p));
+
+prints nothing since `c` is constant.
+
+The `purify ` function also reports mutable private and internal propertes, which are currently not frozen by the `harden` function. 
+
+	class C {
+		#f
+		constructor(f) {
+			this.#f = f;
+		}
+	}
+	const values = {
+		c: new C(0),
+		date: new Date('1993-3-28'),
+		dataView: new DataView(new ArrayBuffer(10)),
+		map: new Map([["a", 0],["b", 1],["c", 2],["d", 2]]),
+		set: new Set([0, 1, 2, 3]),
+		typedArray: new Uint8Array([0, 1, 2, 3]),
+		weakMap: new WeakMap(),
+		weakSet: new WeakSet(),
+	};
+	harden(values);
+	print(purify(values));
+
+prints
+
+	.c.#f
+	.date[[DateValue]]
+	.dataView[[ViewedArrayBuffer]][[ArrayBufferData]]
+	.map[[MapData]]
+	.set[[SetData]]
+	.typedArray[[ViewedArrayBuffer]][[ArrayBufferData]]
+	.weakMap[[WeakMapData]]
+	.weakSet[[WeakSetData]]
+
+##override mistake
+
+I did nothing to workaround the so-called "override mistake”:
+
+	lockdown();
+	const x = harden({foo: 88});
+	const y = Object.create(x);
+	y.foo = 99; // fail
+
+In fact I tend to agree with Allen Wirfs-Brock: [it is not a mistake](https://web.archive.org/web/20141230041441/http://wiki.ecmascript.org/doku.php?id=strawman:fixing_override_mistake).
+
+Even if you use JS as a protoype based language, it is always possible to override:
+
+	Object.defineProperty(y, "foo", { value:99 });
+
+And once you use JS as a class based language, there are no problems:
+
+	lockdown();
+	class C {
+		constructor() {
+		}
+		test() {
+			print("C");
+		}
+	}
+	class D extends harden(C) {
+		constructor() {
+			super();
+		}
+		test() {
+			print("D");
+		}
+	}
+	const d = new D;
+	d.test();
+
+To be discussed, of course!
```

Comment:
Aye, these arguments are strong on theory, but in practice, this is a common gotcha that weakens the claim “Most existing JavaScript modules already work in hardened JavaScript because they apply best practices and avoid mutating shared intrinsics”. Notably, most code generated from TypeScript trips the property override hazard when assigning to Constructor.prototype.constructor, typically as a result of code downcompiled from `class ExtendedError extends Error`, which is common enough to be a pestilence on compatibility.